### PR TITLE
Add PWF Venom support

### LINKED ISSUE #109: Add Venom Support
Add support for Playing With Fusion's Venom Motorcontrollers.

### PR DESCRIPTION
**Pull request by Piphi5**
Fixes #109 

- [x] Add to GUI
- [x] Add to the Robot Code
- [x] Add to the Analyzer

### CODE REVIEW

File: project_utils/cpp/generation/SysIdSetup.cpp [hunk 1]
```diff
@@ -133,6 +142,14 @@ void SetupEncoders(std::string encoderType, bool isEncoding, int period,
                    ->GetSelectedSensorVelocity() /
                cpr / 0.1;  // Conversion factor from 100 ms to seconds
       };
+    } else if (controllerName == "Venom") {
+      position = [=] {
+        return dynamic_cast<frc::CANVenom*>(controller)->GetPosition();
+      };
+      rate = [=] {
+        return dynamic_cast<frc::CANVenom*>(controller)->GetSpeed() /
+               60;  // Needs in rotations per second
```
Comment (calcmogul) [inline project_utils/cpp/generation/SysIdSetup.cpp:151]: "Needs to be in rotations per second"?

Comment (CodingWithFusion) [reply project_utils/cpp/generation/SysIdSetup.cpp:151]: Looks good.  That API returns RPM

**Review by CodingWithFusion — APPROVED** [review]

**Comment by Piphi5** [comment]
@CodingWithFusion just to double-check, there's currently no method to mess with the [number of samples collected](https://www.ctr-electronics.com/downloads/api/cpp/html/classctre_1_1phoenix_1_1sensors_1_1_c_a_n_coder.html#a45881d65a767f78fca83b236dca8b719) and [the period](https://www.ctr-electronics.com/downloads/api/cpp/html/classctre_1_1phoenix_1_1sensors_1_1_c_a_n_coder.html#a5d0cf80979e4705c168836830545d416) in which the samples are collected for velocity?

**Comment by CodingWithFusion** [comment]
Correct, these can't be adjusted
-Jared

File: project_utils/cpp/generation/SysIdSetup.cpp [hunk 2]
```diff
@@ -133,6 +142,14 @@ void SetupEncoders(std::string encoderType, bool isEncoding, int period,
                    ->GetSelectedSensorVelocity() /
                cpr / 0.1;  // Conversion factor from 100 ms to seconds
       };
+    } else if (controllerName == "Venom") {
+      position = [=] {
+        return dynamic_cast<frc::CANVenom*>(controller)->GetPosition();
+      };
+      rate = [=] {
+        return dynamic_cast<frc::CANVenom*>(controller)->GetSpeed() /
+               60;  // Needs to be converted to rotations per second
```
Comment (calcmogul) [inline project_utils/cpp/generation/SysIdSetup.cpp:151]: We should probably include the fact it's converting from RPM to rotations per second.

**Comment by Piphi5** [comment]
@CodingWithFusion what are the filter setup and default settings that are used for the built-in encoder?

**Comment by CodingWithFusion** [comment]
There's no filter per se.   We take the time of the last encoder pulse minus the time of the first encoder pulse divided by the number of encoder pulses we saw each 10 millisecond sample.   So effectively this is averaging as many pulses as we see every 10 milliseconds. There are 4096 encoder counts per revolution.   Im a little rusty.  i'd have to dig out my signals and systems textbook to characterize this any better
-Jared

**Comment by calcmogul** [comment]
OK, so pulses per 10ms. That means the data used for the velocity measurement will, on average, be from 5ms in the past.

**Comment by CodingWithFusion** [comment]
yeah

**Review by prateekma — APPROVED** [review]

**Comment by Piphi5** [comment]
@CodingWithFusion Currently the 2022 [playing with fusion library](https://www.playingwithfusion.com/frc/2022/cppdoc/html/classfrc_1_1_c_a_n_venom.html) inherits from the deprecated SpeedController class, however, SysId is using the new MotorController class meaning that we won't be able to support venom until a version is released that has it inheriting from SpeedController.

**Comment by CodingWithFusion** [comment]
I'll start working on the change to MotorController

**Comment by CodingWithFusion** [comment]
Fixed in libPlayingWithFusion 2022.00.31

**Comment by Piphi5** [comment]
It seems like MotorController is used, but some of the virtual functions arent implemented:
`error: invalid new-expression of abstract class type 'frc::CANVenom'`...
`note:   because the following virtual functions are pure within 'frc::CANVenom': `

I think its specifically:
```/wpilibc-cpp-2022.1.1-beta-4-headers/frc/MotorSafety.h:94:23: note:        virtual std::__cxx11::string frc::MotorSafety::GetDescription() const                                                                            virtual std::string GetDescription() const = 0;```

**Comment by CodingWithFusion** [comment]
Good catch.   Fixed in libPlayingWithFusion 2022.01.03

**Comment by Piphi5** [comment]
```
earched in the following locations:                                                                                                                     - file:/home/piphi5/releases/maven/development/com/playingwithfusion/frc/PlayingWithFusion-driver/2022.01.03/PlayingWithFusion-driver-2022.01.03.pom                                                                                                                                                          - https://frcmaven.wpi.edu/artifactory/development/com/playingwithfusion/frc/PlayingWithFusion-driver/2022.01.03/PlayingWithFusion-driver-2022.01.03.pom                                                                                                                                                      - https://frcmaven.wpi.edu/artifactory/vendor-mvn-release/com/playingwithfusion/frc/PlayingWithFusion-driver/2022.01.03/PlayingWithFusion-driver-2022.01.03.pom                                                                                                                                               - https://www.playingwithfusion.com/frc/maven/com/playingwithfusion/frc/PlayingWithFusion-driver/2022.01.03/PlayingWithFusion-driver-2022.01.03.pom                                                                                                                                                           - https://repo.maven.apache.org/maven2/com/playingwithfusion/frc/PlayingWithFusion-driver/2022.01.03/PlayingWithFusion-driver-2022.01.03.pom 
```

**Comment by CodingWithFusion** [comment]
It was... I jumped the gun.   Try it now.

**Comment by Piphi5** [comment]
Looks like it compiles!

**Comment by Piphi5** [comment]
[PR ](https://github.com/wpilibsuite/sysid/pull/290) has been made

**Pull request merged**